duktape/src/main/jni/JavaScriptObject.cpp

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (C) 2016 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "JavaScriptObject.h"
+#include <stdexcept>
+#include "JString.h"
+#include "JavaExceptions.h"
+
+// TODO: debug instance tracking issues so we don't need to lookup by name on a call.
+#define NO_INSTANCE_VAR
+
+namespace {
+
+// Internal names used for properties in a proxied JavaScript object.
+// The \xff\xff part keeps the variable hidden from JavaScript (visible through C API only).
+const char* WRAPPER_THIS_PROP_NAME = "\xff\xffwrapper_this";
+
+}
+
+JavaScriptObject::JavaScriptObject(JNIEnv* env, duk_context* context, const std::string& name,
+                                   jobjectArray methods)
+    : m_name(name)
+    , m_context(context)
+    , m_instance(nullptr)
+    , m_nextFinalizer(nullptr) {
+  duk_push_global_object(m_context);
+  if (!duk_get_prop_string(m_context, -1, m_name.c_str())) {
+    duk_pop(m_context);
+    throw std::invalid_argument("A global JavaScript object called " + m_name + " was not found");
+  }
+
+#ifdef NO_INSTANCE_VAR
+  if (!duk_is_object(m_context, -1)) {
+#else
+  m_instance = duk_get_heapptr(m_context, -1);
+  if (m_instance == nullptr) {
+#endif
+    duk_pop_2(m_context);
+    throw std::invalid_argument("JavaScript global called " + m_name + " is not an object");
+  }
+
+  if (duk_has_prop_string(m_context, -1, WRAPPER_THIS_PROP_NAME)) {
+    duk_pop_2(m_context);
+    throw std::invalid_argument("JavaScript object " + m_name + " is already proxied");
+  }
+
+  // Make sure that the object has all of the methods we want.
+  jmethodID getName = nullptr;
+  const jsize numArgs = env->GetArrayLength(methods);
+  for (jsize i = 0; i < numArgs; ++i) {
+    auto method = env->GetObjectArrayElement(methods, i);
+    if (getName == nullptr) {
```

Comment:
is this an optimization to avoid looking up the class for java.lang.Method ?  A little bizarre